multidimensional-pod-autoscaler/pkg/recommender/checkpoint/checkpoint_writer.go

```diff
@@ -0,0 +1,152 @@
+	StoreCheckpoints(ctx context.Context, now time.Time, minCheckpoints int) error
+}
+
+type checkpointWriter struct {
+	vpaCheckpointClient vpa_api.VerticalPodAutoscalerCheckpointsGetter
+	cluster             *model.ClusterState
+}
+
+// NewCheckpointWriter returns new instance of a CheckpointWriter
+func NewCheckpointWriter(cluster *model.ClusterState, vpaCheckpointClient vpa_api.VerticalPodAutoscalerCheckpointsGetter) CheckpointWriter {
+	return &checkpointWriter{
+		vpaCheckpointClient: vpaCheckpointClient,
+		cluster:             cluster,
+	}
+}
+
+func isFetchingHistory(mpa *model.Mpa) bool {
+	condition, found := mpa.Conditions[mpa_types.FetchingHistory]
+	if !found {
+		return false
+	}
+	return condition.Status == v1.ConditionTrue
+}
+
+func getMpasToCheckpoint(clusterMpas map[model.MpaID]*model.Mpa) []*model.Mpa {
+	mpas := make([]*model.Mpa, 0, len(clusterMpas))
+	for _, mpa := range clusterMpas {
+		if isFetchingHistory(mpa) {
+			klog.V(3).Infof("MPA %s/%s is loading history, skipping checkpoints", mpa.ID.Namespace, mpa.ID.MpaName)
+			continue
+		}
+		mpas = append(mpas, mpa)
+	}
+	sort.Slice(mpas, func(i, j int) bool {
+		return mpas[i].CheckpointWritten.Before(mpas[j].CheckpointWritten)
+	})
+	return mpas
+}
+
+func (writer *checkpointWriter) StoreCheckpoints(ctx context.Context, now time.Time, minCheckpoints int) error {
+	mpas := getMpasToCheckpoint(writer.cluster.Mpas)
+	for _, mpa := range mpas {
+
+		// Draining ctx.Done() channel. ctx.Err() will be checked if timeout occurred, but minCheckpoints have
+		// to be written before return from this function.
+		select {
+		case <-ctx.Done():
+		default:
+		}
+
+		if ctx.Err() != nil && minCheckpoints <= 0 {
+			return ctx.Err()
+		}
+
+		aggregateContainerStateMap := buildAggregateContainerStateMap(mpa, writer.cluster, now)
+		for container, aggregatedContainerState := range aggregateContainerStateMap {
+			containerCheckpoint, err := aggregatedContainerState.SaveToCheckpoint()
+			if err != nil {
+				klog.Errorf("Cannot serialize checkpoint for mpa %v container %v. Reason: %+v", mpa.ID.MpaName, container, err)
+				continue
+			}
+			checkpointName := fmt.Sprintf("%s-%s", mpa.ID.MpaName, container)
+			vpaCheckpoint := vpa_types.VerticalPodAutoscalerCheckpoint{
+				ObjectMeta: metav1.ObjectMeta{Name: checkpointName},
+				Spec: vpa_types.VerticalPodAutoscalerCheckpointSpec{
+					ContainerName: container,
+					VPAObjectName: mpa.ID.MpaName,
+				},
+				Status: *containerCheckpoint,
+			}
+			err = api_util.CreateOrUpdateMpaCheckpoint(writer.vpaCheckpointClient.VerticalPodAutoscalerCheckpoints(mpa.ID.Namespace), &vpaCheckpoint)
+			if err != nil {
+				klog.Errorf("Cannot save MPA %s/%s checkpoint for %s. Reason: %+v",
+					mpa.ID.Namespace, vpaCheckpoint.Spec.VPAObjectName, vpaCheckpoint.Spec.ContainerName, err)
+			} else {
+				klog.V(3).Infof("Saved MPA %s/%s checkpoint for %s",
+					mpa.ID.Namespace, vpaCheckpoint.Spec.VPAObjectName, vpaCheckpoint.Spec.ContainerName)
```

Comment:
I was actually looking at this myself as well and something I tried to verify but was running into an issue with KinD where I could register the CRD nicely. The MPA was already working in that it scaled up the pod, but maybe it was working incorrectly. Do you think it makes sense to have a separate checkpoint CRD for the MPA? I'm also not sure if other code that is getting imported from the VPA might require the VPA checkpoint to be used.

Reply:
Having a project depend on another project's CRDs may create an awkward dependency between the two projects.

I'm not sure on the best way forward, may be for now it's worth noting that the MPA depends on having the VPA CRDs installed?

Reply:
I was thinking it should be possible to import the type of the VPA checkpoint as the MPA checkpoint and generate it that way. I'll give that a quick try in a moment. If at some point the MPA and VPA start converging more it might make sense to just have the one resource.

Reply:
I've now added the `mpacheckpoint` CRD and should have the code migrated to that. This should allow somewhat cleaner separation, at least from a deployment standpoint.